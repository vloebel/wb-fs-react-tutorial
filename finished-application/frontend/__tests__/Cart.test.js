import { render, screen, waitFor } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';
import userEvent from '@testing-library/user-event';
import wait from 'waait';
import Cart from '../components/Cart';
import { CURRENT_USER_QUERY } from '../components/User';
import { REMOVE_FROM_CART_MUTATION } from '../components/RemoveFromCart';
import { fakeUser, fakeCartItem } from '../lib/testUtils';
import { CartStateProvider } from '../components/LocalState';

const mocks = [
  // {
  //   request: { query: CURRENT_USER_QUERY },
  //   result: {
  //     data: {
  //       authenticatedItem: {
  //         __typename: 'User',
  //         ...fakeUser(),
  //         cart: [fakeCartItem()],
  //       },
  //     },
  //   },
  // },
  {
    request: { query: CURRENT_USER_QUERY },
    result: {
      data: {
        authenticatedItem: {
          ...fakeUser(),
          cart: [fakeCartItem(), fakeCartItem({ id: 'haha123' })],
        },
      },
    },
  },
  {
    request: { query: REMOVE_FROM_CART_MUTATION, variables: { id: 'omg123' } },
    result: {
      data: {
        deleteCartItem: { id: 'omg123' },
      },
    },
  },
];

describe('<Cart/>', () => {
  it('renders and matches snappy', async () => {
    const { container } = render(
      <CartStateProvider>
        <MockedProvider mocks={mocks}>
          <Cart />
        </MockedProvider>
      </CartStateProvider>
    );
    await screen.findByTestId('cart');
    expect(container).toMatchSnapshot();
  });
});

describe('<RemoveFromCart/>', () => {
  it('updates when an item is removed', async () => {
    const { container } = render(
      <CartStateProvider>
        <MockedProvider
          mocks={mocks}
          // defaultOptions={{ watchQuery: { fetchPolicy: 'no-cache' } }}
        >
          <Cart />
        </MockedProvider>
      </CartStateProvider>
    );
    await screen.findByTestId('cart');
    expect(container).toHaveTextContent(/You have 2 items/i);
    const deleteButtons = screen.getAllByTitle(/Remove Item/i);
    expect(deleteButtons).toHaveLength(2);
    userEvent.click(deleteButtons[0]);
    await waitFor(() => wait());
    const deleteButtonsUpdated = screen.getAllByTitle(/Remove Item/i);
    expect(container).toHaveTextContent(/You have 1 item/i);
    expect(deleteButtonsUpdated).toHaveLength(1);
  });
});

export { mocks };
