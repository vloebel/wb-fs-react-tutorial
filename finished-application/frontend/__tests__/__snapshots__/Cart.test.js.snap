// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cart/> renders and matches snappy 1`] = `
<div>
  <div
    class="CartStyles-sc-1tvog53-0 beGvYo"
    data-testid="cart"
  >
    <header>
      <button
        class="CloseButton-sc-1seb878-0 dIMWdj"
        title="close"
      >
        ×
      </button>
      <h3
        class="Supreme-xv30qb-0 hLAPHn"
      >
        Miss Coleman Berge
        's Cart
      </h3>
      <p>
        You Have 
        2
         Item
        s
         in your cart.
      </p>
    </header>
    <ul>
      <li
        class="CartItem__CartItemStyles-sc-1rm9l7o-0 dwvXQK"
      >
        <img
          alt="dogs are best"
          src="dog.jpg"
          width="100"
        />
        <div
          class="cart-item-details"
        >
          <h3>
            dogs are best
          </h3>
          <p>
            $150
             - 
            <em>
              3
               × 
              $50
               each
            </em>
          </p>
        </div>
        <button
          class="RemoveFromCart__BigButton-emvtd6-0 jWbKEf"
          title="Remove Item From Cart"
        >
          ×
        </button>
      </li>
      <li
        class="CartItem__CartItemStyles-sc-1rm9l7o-0 dwvXQK"
      >
        <img
          alt="dogs are best"
          src="dog.jpg"
          width="100"
        />
        <div
          class="cart-item-details"
        >
          <h3>
            dogs are best
          </h3>
          <p>
            $150
             - 
            <em>
              3
               × 
              $50
               each
            </em>
          </p>
        </div>
        <button
          class="RemoveFromCart__BigButton-emvtd6-0 jWbKEf"
          title="Remove Item From Cart"
        >
          ×
        </button>
      </li>
    </ul>
    <footer>
      <p>
        $300
      </p>
      <form
        class="Checkout__CheckoutFormStyles-sc-10mtoy5-0 eNIfuo"
        data-testid="checkout"
      >
        <div />
        <button
          class="SickButton-l04z44-0 ievaJh"
          type="submit"
        >
          Pay
        </button>
      </form>
    </footer>
  </div>
</div>
`;
